### PR TITLE
Use pylint

### PR DESCRIPTION
**Pull request by akuzminsky**
* Pylint fixes
* Gzip modifier as a part of refactoring
* other refactoring
* More docs, etc

### CODE REVIEW

File: twindb_backup/__init__.py [hunk 1]
```diff
@@ -61,36 +96,64 @@ def setup_logging(logger, debug=False):     # pragma: no cover
...
     if keep_copies == 0:
         return all_files
     else:
         return all_files[:-keep_copies]
 
 
 def delete_local_files(dir_backups, keep_copies):
+    """Deletes local backup copies based on given retention number.
+
+    :param dir_backups: directory with backup copies
```
Comment (ovaistariq) [inline twindb_backup/__init__.py]: param type is missing

File: twindb_backup/__init__.py [hunk 2]
```diff
@@ -61,36 +96,64 @@ def setup_logging(logger, debug=False):     # pragma: no cover
...
     return backup_dirs
 
 
 def get_files_to_delete(all_files, keep_copies):
-    log.debug('Retain %d files', keep_copies)
+    """If you give it a list of files and number of how many
+    you'd like to keep the function will return files that need
+    to be deleted
+
+    :param all_files: list of strings
```
Comment (ovaistariq) [inline twindb_backup/__init__.py]: param type is missing

File: twindb_backup/__init__.py [hunk 3]
```diff
@@ -61,36 +96,64 @@ def setup_logging(logger, debug=False):     # pragma: no cover
 
 
 def get_directories_to_backup(config):
+    """Get directories to backup from a config file
+
+    :param config: instance of ConfigParser()
```
Comment (ovaistariq) [inline twindb_backup/__init__.py]: param type is missing

File: twindb_backup/destination/s3.py [hunk 4]
```diff
@@ -37,21 +45,43 @@
 
 
 class S3Error(DestinationError):
+    """S3 destination errors"""
     pass
 
 
+class AWSAuthOptions(object):  # pylint: disable=too-few-public-methods
```
Comment (ovaistariq) [inline twindb_backup/destination/s3.py]: What is the benefit of this separate class? I think this indirection has no benefit.

File: twindb_backup/destination/s3.py [hunk 5]
```diff
@@ -207,39 +242,35 @@ def get_stream(self, path):
...
             # The read end of the pipe must be closed in the child process
             # before we start writing to it.
             os.close(read_fd)
 
-            with os.fdopen(write_fd, 'wb') as write_pipe:
-                s3_client.download_fileobj(bucket_name, key, write_pipe)
+            with os.fdopen(write_fd, 'wb') as w_pipe:
+                s3_client.download_fileobj(bucket_name, key, w_pipe)
 
-        try:
```
Comment (ovaistariq) [inline twindb_backup/destination/s3.py]: You should still keep the `try` block here and remove the `except` block. So that you are sure to call `download_proc.start` in the `finally` block at all costs.

File: twindb_backup/destination/ssh.py [hunk 6]
```diff
@@ -1,31 +1,56 @@
...
-from twindb_backup import log
+
+from twindb_backup import LOG
 from twindb_backup.destination.base_destination import BaseDestination, \
     DestinationError
+from twindb_backup.util import run_command
 
 
-class Ssh(BaseDestination):
+class SshConnectInfo(object):  # pylint: disable=too-few-public-methods
```
Comment (ovaistariq) [inline twindb_backup/destination/ssh.py]: What is the benefit of having this as a separate class if this is only ever used in the `Ssh` class.

Comment (akuzminsky) [inline twindb_backup/destination/ssh.py]: Linter was complaining about too many arguments in the function.

File: twindb_backup/destination/local.py [hunk 7]
```diff
@@ -45,51 +53,33 @@ def list_files(self, prefix, recursive=False):
...
-        log.debug('Running %s', ' '.join(cmd))
-        proc = Popen(cmd, stdout=PIPE, stderr=PIPE)
-        cout, cerr = proc.communicate()
 
-        return sorted(cout.split())
+        with run_command(cmd) as cout:
+            return sorted(cout.split())
 
-    def find_files(self, prefix):
+    def find_files(self, prefix, run_type):
```
Comment (ovaistariq) [inline twindb_backup/destination/local.py]: I don't see `run_type` being used. Is it needed here?

File: twindb_backup/destination/ssh.py (hunk 6 above)
Comment (ovaistariq) [inline twindb_backup/destination/ssh.py]: OK we should make it a schematics type in a future release.

File: twindb_backup/source/mysql_source.py [hunk 8]
```diff
@@ -1,32 +1,101 @@
...
+
+        :param mysql_connect_info: MySQL connection details
+        :type mysql_connect_info: MySQLConnectInfo
+        :param run_type:
+        :param full_backup: When to do full backup e.g. daily, weekly
+        :type full_backup: str
+        :param dst:
+        """
+
+        class _BackupInfo(object):  # pylint: disable=too-few-public-methods
```
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: We should set this as a schematics type in the next diff. Though its trivial to do it even in this one.
Why is this class defined inside the `init` function? I would think that this would be usable outside of the class as well. Or may be further down in the class, so doesn't make sense to define it inside `init`.

File: twindb_backup/source/mysql_source.py [hunk 9]
```diff
@@ -188,44 +277,63 @@ def _get_backup_type(self):
 
     @property
     def type(self):
+        """Get backup copy type - full or incremental
+
+        :return: 'full' or 'incrmental'
+        :rtype: str
+        """
         return self._get_backup_type()
```
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: Why not move the code from _get_backup_type to this getter property.

File: twindb_backup/source/mysql_source.py [hunk 10]
```diff
@@ -188,44 +277,63 @@ def _get_backup_type(self):
...
+        or its own name if the copy is a full copy.
+        :rtype: str
+        """
+        return sorted(self.status[self.full_backup].keys(), reverse=True)[0]
 
     @property
     def parent_lsn(self):
-        full_backup = self.config.get('mysql', 'full_backup')
-        return self.status[full_backup][self.parent]['lsn']
+        """LST of the parent backup copy.
```
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: Typo

File: twindb_backup/source/mysql_source.py [hunk 11]
```diff
@@ -188,44 +277,63 @@ def _get_backup_type(self):
...
+
+        :return: LSN of the parent or its own LSN
+        if the backup copy is a full copy.
+        """
+        return self.status[self.full_backup][self.parent]['lsn']
 
     def _parent_exists(self):
```
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: Shouldn't this function make use of the `parent` property?

File: twindb_backup/__init__.py (hunk 2 above)
Comment (akuzminsky) [inline twindb_backup/__init__.py]: fixed.

File: twindb_backup/__init__.py (hunk 1 above)
Comment (akuzminsky) [inline twindb_backup/__init__.py]: fixed

File: twindb_backup/destination/local.py (hunk 7 above)
Comment (akuzminsky) [inline twindb_backup/destination/local.py]: I don't like this function at all. It's not clear what it does and how it's different from list_files().
I suggest redesign this part in another PR.

File: twindb_backup/source/mysql_source.py (hunk 8 above)
Comment (akuzminsky) [inline twindb_backup/source/mysql_source.py]: BackupInfo should describe a status entry. At this moment it's not used outside init(), but you're right, we should check where else it may be used.
```
"master1/hourly/mysql/mysql-2017-02-13_04_58_53.xbstream.gz": {
            "binlog": "mysql-bin.000001",
            "config": [
                {
                    "/etc/my.cnf": "W215c3FsZF0KZGF0YWRpcj0vdmFyL2xpYi9teXNxbApzb2NrZXQ9L3Zhci9saWIvbXlzcWwvbXlzcWwuc29jawp1c2VyPW15c3FsCiMgRGlzYWJsaW5nIHN5bWJvbGljLWxpbmtzIGlzIHJlY29tbWVuZGVkIHRvIHByZXZlbnQgYXNzb3J0ZWQgc2VjdXJpdHkgcmlza3MKc3ltYm9saWMtbGlua3M9MAoKc2VydmVyX2lkPTEwMApndGlkX21vZGU9T04KbG9nLWJpbj1teXNxbC1iaW4KbG9nLXNsYXZlLXVwZGF0ZXMKZW5mb3JjZS1ndGlkLWNvbnNpc3RlbmN5CgpbbXlzcWxkX3NhZmVdCmxvZy1lcnJvcj0vdmFyL2xvZy9teXNxbGQubG9nCnBpZC1maWxlPS92YXIvcnVuL215c3FsZC9teXNxbGQucGlkCg=="
                }
            ],
            "lsn": 19621635,
            "position": 40713,
            "type": "full"
        }
```

File: twindb_backup/source/mysql_source.py (hunk 9 above)
Comment (akuzminsky) [inline twindb_backup/source/mysql_source.py]: Oh, this is where _BackupInfo() may be useful

File: twindb_backup/source/mysql_source.py (hunk 11 above)
Comment (akuzminsky) [inline twindb_backup/source/mysql_source.py]: It's a little bit confusing, I'll rename the function.
It does tell if an eligible full copy exists so `twindb-backup` can take incremental copy.

File: twindb_backup/destination/local.py (hunk 7 above)
Comment (ovaistariq) [inline twindb_backup/destination/local.py]: OK, but doesn't the linter complain about unused parameter `run_type`? That would be surprising if it doesn't do that.

File: twindb_backup/source/mysql_source.py (hunk 8 above)
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: Could you move it out of here :)

File: twindb_backup/source/mysql_source.py (hunk 9 above)
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: Yes. And why do you have `_get_backup_type` separately?

File: twindb_backup/source/mysql_source.py (hunk 10 above)
Comment (ovaistariq) [inline twindb_backup/source/mysql_source.py]: LSN

File: twindb_backup/destination/local.py (hunk 7 above)
Comment (akuzminsky) [inline twindb_backup/destination/local.py]: It doesn't because run_type is an argument of the parent method. Actually, it complains when I remove it.

File: twindb_backup/source/mysql_source.py (hunk 8 above)
Comment (akuzminsky) [inline twindb_backup/source/mysql_source.py]: Now? Let's do it in a separate PR.

File: twindb_backup/source/mysql_source.py (hunk 9 above)
Comment (akuzminsky) [inline twindb_backup/source/mysql_source.py]: This is a function that tell which backup type the tool should take now
